pythonFiles/include/blender_vscode/installation.py

```diff
@@ -1,55 +1,93 @@
 
+    for package in package_names:
+        if not is_module_installed(package):
+            handle_fatal_error(
+                textwrap.dedent(
+                    f"""\
+            Package can not be imported: {package}
+            If you are experiencing PermissionError please scroll up to the line `Execute:` and execute the command manually.
+            Also close all Blender instances.
+            For old system wide Blender installations you might to reinstall Blender."""
+                )
+            )
 
-def packages_are_installed(package_names):
-    return all(module_can_be_imported(name) for name in package_names)
 
-
-def install_packages(package_names):
-    if not module_can_be_imported("pip"):
+def ensure_packages_are_installed(package_names: List[str]):
+    if not is_module_installed("pip"):
         install_pip()
 
-    for name in package_names:
-        ensure_package_is_installed(name)
-
-    assert packages_are_installed(package_names)
-
-
-def ensure_package_is_installed(name: str):
-    if not module_can_be_imported(name):
-        install_package(name)
-
+    if not any(is_module_installed(name) for name in package_names):
+        print(f"INFO: installing dependencies: {package_names}")
+        install_packages(package_names)
+        return
 
-def install_package(name: str):
+    # Perform precise checks if any dependency changes are needed.
+    # Manual checks prevent from unnecessary updates which on <0.21 installations may require admin rights
+    # Moreover upgrading when using --target is not possible.
+    requires_reinstall = False
+    for name in package_names:
+        _name, requested_version = _split_package_version(name)
+        if requested_version is None:
+            continue
+        requested_version = Version.parse(requested_version)
+        real_version = Version.parse(package_version(name))
+        assert isinstance(requested_version, Version), requested_version
+        assert isinstance(real_version, Version), real_version
+        if "==" in name and not (real_version == requested_version):
+            requires_reinstall = f"{name}, got {real_version}"
+            break
+        elif "<=" in name and not (real_version <= requested_version):
+            requires_reinstall = f"{name}, got {real_version}"
+            break
+        elif "<=" not in name and "<" in name and not (real_version < requested_version):
+            requires_reinstall = f"{name}, got {real_version}"
+            break
+        elif ">=" not in name and ">" in name and not (real_version > requested_version):
+            requires_reinstall = f"{name}, got {real_version}"
+            break
+
+    if requires_reinstall:
+        print(f"INFO: dependencies reuqire update because of {requires_reinstall}, reinstalling...")
+        print(f'INFO: removing "{bpy.utils.user_resource("SCRIPTS", path="modules")}"')
+        shutil.rmtree(bpy.utils.user_resource("SCRIPTS", path="modules"))
```

Comment:
This carries a low risk of removing user data... The only think that can be done is to move the dir to backup location or track exactly what files are being installed.

Reply:
What makes you think that the risk is low? Sounds like removing a folder that is not specific to the extension is quite risky.
May be better to tell the user in some way that the folder should be removed instead of just doing it.

Reply:
i don't think this folder is used often. 
The alternative is to track all names of dependencies and remove folder and .dist-info folder. Like `jinja` and `^jinja.*-.*\.dist-info$`. A static list of dependencies is easiest to write.

Reply:
mad a quick implementation, might still have mistakes

Some snipet with alternative:
```
def list_installed_packages():
    target = get_package_install_directory()
    command = [str(python_path), "-m", "pip", "freeze", "--disable-pip-version-check", "--exclude-editable", "---path", target]
    print("INFO: execute: ", " ".join(_escape_space(c) for c in command))
    output = subprocess.check_output(command, cwd=_CWD_FOR_SUBPROCESSES).decode()
    for package in output.splitlines():
        yield package.split("==")[-1]
```